Add OktaDev Schematics post (#1262)
oktadeveloper/okta-blog: _source/_posts/2022-10-01-quick-javascript-login.adoc

---
layout: blog_post
title: "Quick JavaScript Login with Schematics"
author: matt-raible
by: advocate
communities: [javascript]
description: ""
tags: []
tweets:
- ""
- ""
- ""
image:
type: conversion
---
:page-liquid:
:toc: macro
:experimental:

// Other title ideas:

> //   - Quick JavaScript Login with Schematics (current, 58)

Reviewer: How about:

Quickstart JavaScript Authentication with OktaDev Schematics (60)

I'm also in favor of the current one, it makes the most sense for someone who doesn't even know about schematics.

Author: @havi Do you think one is better than the other?

Author: I'm going to change it to `Quick JavaScript Authentication with OktaDev Schematics`. It gets a 62. Thanks for the suggestion!

//   - Schematics for Quick JavaScript Authentication (59)
//   - Quick Templates for JavaScript Authentication (59)
//   - CLI to App to Auth in Minutes with OktaDev Schematics! (60)

// Keywords:
//   - javascript authentication: 480 (there's an Okta post on the first page)
//   - javascript auth: 210
//   - javascript login: 880
//   - javascript security: 720
//   - javascript frameworks: 27,100
//   - react oidc: 720
//   - oidc authentication: 2900
//   - jwt auth: 2400
//   - jwt javascript: 720

// Should we emphasis this is a developer story moreso than a product pitch?

Hello,

My name is Matt Raible, and I am a developer advocate at Okta. I've worked here for over five years; before that I was an independent consultant for 20 years, doing mostly Java and web development for clients.

> I've learned a lot about OpenID Connect (OIDC) and Okta's SDKs over my Okta years and only recently tried to use Auth0's. Over the last couple of weeks, I integrated many Auth0 SDKs into OktaDev Schematics.

Reviewer: This paragraph is a bit hard to follow, and to understand why it is here without knowing what OktaDev Schematics is. I like that you are trying to add some background, but perhaps we can connect it better to the paragraph below, or make a previous reference to what OktaDev Schematics is.

Author: I updated. Please have another look at this section.

type: conversion
---
:page-liquid:
:toc: macro
:experimental:

// Other title ideas:
//   - Quick JavaScript Login with Schematics (current, 58)
//   - Schematics for Quick JavaScript Authentication (59)
//   - Quick Templates for JavaScript Authentication (59)
//   - CLI to App to Auth in Minutes with OktaDev Schematics! (60)

// Keywords:
//   - javascript authentication: 480 (there's an Okta post on the first page)
//   - javascript auth: 210
//   - javascript login: 880
//   - javascript security: 720
//   - javascript frameworks: 27,100
//   - react oidc: 720
//   - oidc authentication: 2900
//   - jwt auth: 2400
//   - jwt javascript: 720

// Should we emphasize this is a developer story moreso than a product pitch?

Hello,

My name is Matt Raible, and I am a developer advocate at Okta. I've worked here for over five years; before that, I was an independent consultant for 20 years, mainly doing Java and web development for clients.

I've learned a lot about OpenID Connect (OIDC) and Okta's SDKs over my Okta years and only recently tried to use Auth0's. Over the last couple of weeks, I integrated many Auth0 SDKs into OktaDev Schematics.

> I was pleasantly surprised by the experience because the configuration often boils down to two values: `domain` and `clientId`. And then, it just works! There's no need to worry about issuers, redirect URIs, or callback URLs.

Reviewer: When I did my set up I had to configure callback URLs and I had to specify the issuer (tenant). Even though it is true the process was very easy, I still had to configure and capture some of those values.

Author: Good point. I'll revise.

> I was pleasantly surprised by the experience because the configuration often boils down to two values: `domain` and `clientId`. And then, it just works! There's no need to worry about issuers, redirect URIs, or callback URLs.

Author: ```suggestion
I was pleasantly surprised by the experience because the configuration often boils down to two main values: `domain` and `clientId`. You do have to add callback URLs too, but these are often the same as your app's homepage, so you don't need to add routes in your application. And then, it just works! There's no need to worry about issuers or other OIDC concepts. 
```

> This is a strong departure from how we've done things in Okta's SDKs, where we expect people to know about OpenID Connect. And from a developer advocacy standpoint, we often try to educate developers about OAuth and OIDC. Since these topics are complex, they can be challenging to understand. I like the approach of just needing two values, and everything just works!

Reviewer: I'll ask someone else to review this paragraph. It has some negative connotation about Okta, to which I'm not 100% sure about, perhaps it can be rewritten.

Reviewer 2: I agree with @bajcmartinez, it suggests that Okta SDKs don’t do their job well. Even if it's the case, I don't think it's a good thing to admit it publicly 😂 
Maybe it could be reformulated as follows:

"We often try to educate developers about OAuth and OIDC but since these topics are complex, they can be challenging to understand. I like the approach of just needing two values, and everything just works!"

What do you think?

Author: I like it! I've added this as a suggestion.

> This is a strong departure from how we've done things in Okta's SDKs, where we expect people to know about OpenID Connect. And from a developer advocacy standpoint, we often try to educate developers about OAuth and OIDC. Since these topics are complex, they can be challenging to understand. I like the approach of just needing two values, and everything just works!

Author: ```suggestion
This is a strong departure from how we've done things in Okta's SDKs, where we expect people to know about OpenID Connect. We often try to educate developers about OAuth and OIDC but since these topics are complex, they can be challenging to understand. I like the approach of just needing two values, and everything just works!
```

toc::[]

== What is OktaDev Schematics?

> https://github.com/oktadev/schematics[OktaDev Schematics] is a tool I developed to help streamline integrating OIDC authentication into JavaScript (or TypeScript) applications. In my first few years at Okta, I'd demo how to use our Angular SDK at conferences, and I'd inevitably miss a step, causing the demo to fail.

Reviewer: ```suggestion
https://github.com/oktadev/schematics[OktaDev Schematics] is a tool I developed to help streamline the integration of OpenID Connect into JavaScript (or TypeScript) applications.

In my first few years at Okta, I'd demo how to use our Angular SDK at conferences, and I'd inevitably miss a step, causing the demo to fail. A failing demo is frustrating, so I created OktaDev Schematics to fix it.

My idea was simple:
```

Author: @bajcmartinez Your suggestion is probably a good idea. However, I do like to include why I created OktaDev Schematics. Maybe it's not really relevant?

Reviewer: I didn't remove the why, just have it on a separate paragraph.

Author: OK, I'll commit this change shortly.

> A failing demo is frustrating, so I created OktaDev Schematics to fix it. My idea was simple:

Reviewer: ```suggestion
```

1. Create an app using the framework's CLI
2. Integrate OIDC authentication using templates
3. Add a unit test to prove nothing breaks

> The project builds on https://angular.io/guide/schematics[Angular Schematics], so it's incredibly easy to use in Angular projects.

Reviewer: ```suggestion
The project builds on https://angular.io/guide/schematics[Angular Schematics], so it's incredibly easy to use, particularly in Angular projects, but applies to other frameworks as well.
```

[source,shell]
----
ng add @oktadev/schematics
----

> This command will prompt you for an issuer and client ID, and then OktaDev Schematics will configure your app to use them. If you're using the https://github.com/auth0/auth0-cli[Auth0 CLI], you'll need to create a Single-Page Application (SPA) app to get these values.

Reviewer: - Since I'm relatively new to Okta and Auth0, I had issues understanding what the issuer is, until I realized it is the tenant domain in the case of Auth0, correct? If so, can we perhaps make this a bit more explicit?

- You name here the Auth0 cli, which is great, but it can also be done thorugh the web interface. You could opt to explain how to do it with the CLI or add a reference, but we shouldn't suggest that you only use the CLI. And if I'm not mistaking, you can also set this up for an Okta account, right?

Author: Yes, the issuer is `https://<your-auth0-domain>/`, and it needs a trailing slash with some libraries. OktaDev Schematics is smart enough to allow you to just type in the domain too. https://github.com/oktadev/schematics/blob/main/src/add-auth/index.ts#L187

Author: I added a sentence below about the Auth0 dashboard. I'm not sure if we should link to it or not.

[source,shell]
----
auth0 apps create
----

> 

Author: ```suggestion
You can also use the Auth0 dashboard through a browser if you prefer GUIs over CLIs. 

```

Author: I changed it to:

```
TIP: You can also create a SPA app using https:///manage.auth0.com[the Auth0 dashboard] if you prefer GUIs over CLIs.
```

> In the latest release, I've made it so that adding an `--auth0` flag will configure Auth0 instead of Okta.

Reviewer: ```suggestion
In the latest release, I've made it so that adding an `--auth0` flag will set up your project to use Auth0 instead of Okta.
```

Reviewer: The tool does not change any settings on your account, it only updates the local project, correct?

Author: Correct. It doesn't communicate with Okta or Auth0 in any way.

// https://twitter.com/mraible/status/1566907778161049600
++++
{% twitter 1566907778161049600 %}
++++

If you're using Angular or Ionic with Angular, you can create an app and then add OIDC authentication in just two commands!

- Angular: `ng new` and `ng add @oktadev/schematics --auth0`
- Ionic: `ionic start` and `ng add @oktadev/schematics --auth0`

> 

Reviewer: [here]: Paragraph about more information on how to integrate OpenID connect with Angular and Auth0

Author: Do you think I could get away with just linking to my previous post that explains all of this? https://auth0.com/blog/add-oidc-to-angular-apps-quickly/

Reviewer: @mraible that should be fine

The tool is smart enough to detect the framework you're using and if you're using TypeScript or JavaScript.

> If you want to create a React app, you'll need to perform a few steps:

Reviewer: If you want, you could here generalize the steps for react, or vue, as they are pretty much the same ignoring the first command you use to create the project.

Author: Good idea. Will do.

1. Create your React app with Create React App:

>   npx create-react-app

Author: ```suggestion
  npx create-react-app secure-app
```

+
TIP: Add `--template typescript` if you want TypeScript support.

2. Install the Schematics CLI globally:

  npm i -g @angular-devkit/schematics-cli

> 3. Install OktaDev Schematics and run the `add-auth` schematic:

Author: ```suggestion
3. In the `secure-app` project, install OktaDev Schematics and run the `add-auth` schematic:
```

  npm i -D @oktadev/schematics

>   schematics @oktadev/schematics:add-auth

Reviewer: Just to note here, as I'm not sure why it was the case, but I also had to install 'npm-check-updates' before running the schematics command.

Without manually installing that package into the project, I'd get the following error:
Error: Cannot find module 'npm-check-updates'

Author: I think you found a bug! This PR should fix it. https://github.com/oktadev/schematics/pull/957

Author: @bajcmartinez Can you please try again? I just released [v6.0.1](https://github.com/oktadev/schematics/releases/tag/v6.0.1).

Reviewer: Just a question, on NPM, v 6.0.1 is not yet available.

Reviewer: @mraible other than that, looks good now!

Author: Doh! That's because I forgot to update the version before doing the release. Thanks for catching it. Will fix.

As you can guess, the integration is pretty bare-bones. It configures the SDK and adds a login/logout button. I believe this is in line with the apps each CLI generates. They're pretty bare-bones too!

> If you'd like to see how to add user information to an Angular app, please read my https://auth0.com/blog/add-oidc-to-angular-apps-quickly/[Add OpenID Connect to Angular Apps Quickly] tutorial.

Reviewer: Since we moved out of Angular, to react or vue, this paragraph would make more sense above, together with the instruction for Angular.

> If you'd like to see how the templates differ between Okta and Auth0, you can https://github.com/oktadev/schematics/tree/main/src/add-auth[view the templates on GitHub]. If you have a GitHub account, you can also https://github.dev/oktadev/schematics[launch github.dev/oktadev/schematics] and navigate to `src/add-auth/{auth0,okta}`.

Reviewer: ```suggestion
This project is open-source, and the full code, including the templates for both Okta and Auth0 are available on https://github.dev/oktadev/schematics[GitHub].
```

Author: Do you think it's a bad idea to link to the templates directly? That's why I had the previous text. However, I can see how smart developers might be able to figure out where the templates are.

Reviewer: I think the repo link is enough, but if you want link the templates as well, it doesn't hurt, just keep it simple,

image::{% asset_path 'blog/oktadev-schematics-auth0/github.dev.jpg' %}[alt=github.dev is pretty slick!,width=800,align=center]

== JavaScript Framework Support

There's a wealth of popular frameworks currently supported by OktaDev Schematics:

- https://github.com/oktadev/schematics#angular[Angular]
- https://github.com/oktadev/schematics#react[React]
- https://github.com/oktadev/schematics#vue[Vue]

> - https://github.com/oktadev/schematics#ionic[Ionic]

Reviewer: ```suggestion
- https://github.com/oktadev/schematics#ionic[Ionic with Angular]
```

- https://github.com/oktadev/schematics#react-native[React Native]
- https://github.com/oktadev/schematics#express[Express]

See the project's https://github.com/oktadev/schematics#links[links section] if you want to learn more about Okta or Auth0's SDKs.

> The support for Ionic currently only includes Angular. However, Ionic does support React and Vue too. To make OIDC authentication work, I leveraged https://github.com/wi3land/ionic-appauth[Ionic AppAuth]. One cool thing about this library is it has https://github.com/wi3land/ionic-appauth/tree/master/demos/react[React] and https://github.com/wi3land/ionic-appauth/tree/master/demos/vue[Vue] examples available. I recently updated these to work with Capacitor 4.

Reviewer: Not sure how relevant this paragraph is for the article. Are those examples important for the use of OktaDev Schematics? Or do they only belong to the other project?

Author: The examples are important in that they show that Ionic + React or Vue is possible. I tested the examples with both Okta and Auth0 to make sure they work. In the future, it'd probably be a good idea to include them in OktaDev Schematics, but I didn't have the bandwidth for the 6.0 release. Do you think I should mention this?

Reviewer: If you plan on adding those in future versions, you can say that, and in the meanwhile they can check out those links.

> // https://twitter.com/mraible/status/1566829762026291200

Reviewer: Linked to the paragraph above

++++
{% twitter 1566829762026291200 %}
++++

== The Beauty of Continuous Integration

> As part of creating this library, I wrote many tests, including integration tests that:

Reviewer: ```suggestion
To avoid situations where a sample integration would fail in public, like the story I told before, I made sure the library is well tested, limiting the scope of the tests not only to the library code, but also to cover the workflow and generated code bases, e.g.:
```

Author: I added your suggestion. Thanks!

> Create an app with a CLI

Reviewer: ```suggestion
- Create an app with a CLI
```

> Integrate Okta/Auth0 SDKs

Reviewer: ```suggestion
- Integrate Okta/Auth0 SDKs
```

Author: Good catch on the missing bullet points! Fixed.

> Run tests to ensure nothing breaks

Reviewer: ```suggestion
- Validate the integration to make sure nothing breaks
```

Author: Added. Thanks.

These commands are all encapsulated in a https://github.com/oktadev/schematics/blob/main/test-app.sh[`test-app.sh` script] I wrote. It's kinda ugly and can probably be improved, but it works pretty well for now.

If nothing else, this script shows you how to automate creating apps with framework CLIs for continuous integration. The project's https://github.com/oktadev/schematics/blob/main/.github/workflows/apps.yml[Apps workflow] shows how all these frameworks are tested in a matrix.

You might notice that all the CLIs are installed without a version number:

[source,yaml]
----
run: npm i -g @angular-devkit/schematics-cli @angular/cli @vue/cli @ionic/cli
----

This is done on purpose, so these templates are continuously tested against the latest release. There's a https://github.com/oktadev/schematics/blob/main/.github/workflows/nightly.yml[Nightly worflow] that ensures nothing breaks every evening. Dependabot is also https://github.com/oktadev/schematics/blob/main/.github/dependabot.yml[configured] to keep the project's dependencies and SDK versions up-to-date.

== Learn more about Authentication in JavaScript

> Creating and maintaining OktaDev Schematics has greatly benefited me and my knowledge of Okta and Auth0's SDKs. I sometimes find issues with new major releases of frameworks and our SDKs. Since the CLIs always use the latest versions, I often find problems before anyone else. Ahhh, the beauty of testing automation! 😃

Reviewer: ```suggestion
When framework or libraries introduce changes, sometimes things break. But now, thanks to automated tests, I often find problems as soon as new versions are released. What a beauty!
```

Author: I'm not sure that "What a beauty" makes sense.

Reviewer: We can remove that, I added it as you mentioned "The beauty of testing automation".

Author: Changed it to:  Cheers to continuous integration! 😃
